File: LinkedList.rb
class Node
    attr_accessor :item, :next
    def initialize; end

    def initialize(item)
        @item = item.to_i
    end
end

#####

class LinkedList
    @first = Node
    def initialize; end

    ####

    def addItemAtEnd(item)
        if item != Node
            newNode = Node.new(item)
        elsif item == Node
            newNode = item
        else
            raise 'Invalid data type'
        end

        current = @first
        past = nil
        until current.nil?
            past = current
            current = current.next
        end
        if past.nil?
            @first = newNode
        else
            past.next = newNode
            newNode.next = current
        end
    end

    ####

    def addItemAtFirst(item)
        if item != Node
            newNode = Node.new(item)
        elsif item == Node
            newNode = item
        else
            raise 'Invalid data type'
        end
        if @first.nil?
            @first = newNode
        else
            newNode.next = @first
            @first = newNode
        end
   end

    ####

    def addItemToSortedList(item)
        if item != Node
            newNode = Node.new(item)
        elsif item == Node
            newNode = item
        else
            raise 'Invalid data type'
        end
        current = @first
        past = nil
        while !current.nil? && current.item < newNode.item
            past = current
            current = current.next
        end
        if past.nil?
            newNode.next = @first
            @first = newNode
        else
            past.next = newNode
            newNode.next = current
        end
      end

    ####

    def findItem(item)
        current = @first
        current = current.next while !current.nil? && current.item < item
        if current.nil? || current.item > item
            puts 'Item not found'
        else
            puts 'Item found'
        end
    end

    ####

    def deleteItem(item)
        current = @first
        past = nil
        while !current.nil? && current.item != item
            past = current
            current = current.next
        end
        puts 'Item not found' if current.nil? || current.item != item
        if past.nil?
            @first = current.next
        else
            past.next = current.next
        end
    end

    ####

    def reverse
        unless @first.nil?
            current = @first
            past = nil
            post = @first.next
            until current.next.nil?
                current.next = past
                past = current
                current = post
                post = post.next
            end
            current.next = past
            @first = current
        end
    end

    ####

    def printList
        current = @first
        until current.nil?
            puts current.item.to_i
            current = current.next
        end
    end
end

# LinkedListMain
ls = LinkedList .new
ls.addItemAtEnd(1)
ls.addItemAtEnd(5)
ls.addItemAtEnd(4)
ls.addItemAtEnd(3)
ls.addItemAtEnd(2)
ls.addItemAtFirst(18)
ls.addItemAtFirst(19)
ls.addItemToSortedList(84)
ls.addItemToSortedList(7)
ls.addItemToSortedList(2)
ls.addItemToSortedList(4)
ls.addItemToSortedList(7)
ls.addItemToSortedList(1)
ls.deleteItem(18)
ls.deleteItem(2)
ls.findItem(4)
ls.printList

File: LinkedList.java
/**
 * Created by mikdev on 2/12/17.
 */
class LinkedListMain {
    public static void main(String[] args) {
        LinkedList ls = new LinkedList();
        ls.AddItemAtEnd(1);
        ls.AddItemAtEnd(2);
        ls.AddItemAtEnd(3);
        ls.AddItemAtEnd(4);
        ls.AddItemAtEnd(5);
        ls.AddItemAtEnd(6);
        ls.AddItemAtEnd(7);
        ls.AddItemAtFirst(0);
        ls.AddItemAtFirst(-1);
        ls.DeleteItem(-1);
        ls.DeleteItem(4);
        ls.DeleteItem(7);
        ls.PrintList();
        ls.Reverse();
        ls.PrintList();
    }
}

class LinkedList {
    Node first;

    public boolean AddItemAtEnd(int item) {
        Node NewNode = new Node(item);
        AddItemAtEnd(NewNode);
        return true;
    }

    public boolean AddItemAtEnd(Node NewNode) {
        Node Current = first, Past = null;
        while (Current != null) {
            Past = Current;
            Current = Current.next;
        }
        if (Past == null) {
            first = NewNode;
        } else {
            Past.next = NewNode;
            NewNode.next = Current;
        }
        return true;
    }

    public boolean AddItemAtFirst(int item) {
        Node NewNode = new Node(item);
        AddItemAtFirst(NewNode);
        return true;
    }

    public boolean AddItemAtFirst(Node NewNode) {
        if (first != null) {
            NewNode.next = first;
            first = NewNode;
        } else {
            first = NewNode;
        }
        return true;
    }

    public boolean AddItemToSortedList(Node NewNode) {
        Node Current = first, Past = null;
        while (Current != null && Current.item < NewNode.item) {
            Past = Current;
            Current = Current.next;
        }
        if (Past == null) {
            NewNode.next = first;
            first = NewNode;
        } else {
            Past.next = NewNode;
            NewNode.next = Current;
        }
        return true;
    }

    public boolean AddItemToSortedList(int item) {
        Node NewNode = new Node(item);
        AddItemToSortedList(NewNode);
        return true;
    }

    public boolean FindItem(int item) {
        Node Current = first;
        while (Current != null && Current.item < item) {
            Current = Current.next;
        }
        if (Current == null || Current.item > item)
            return false;
        else
            System.out.println("Item found");

        return true;
    }

    public void Reverse() {
        if (first != null) {
            Node Past = null, Current = first, Next = first.next;

            while (Current.next != null) {
                Current.next = Past;
                Past = Current;
                Current = Next;
                Next = Next.next;
            }
            Current.next = Past;
            first = Current;
        }
    }

    public boolean DeleteItem(Node NewNode) {
        DeleteItem(NewNode.item);
        return true;
    }

    public boolean DeleteItem(int item) {
        Node Current = first, Past = null;
        while (Current != null && Current.item != item) {
            Past = Current;
            Current = Current.next;
        }
        if (Current == null || Current.item != item) {
            return false;
        }
        if (Past == null) {
            first = Current.next;
        } else {
            Past.next = Current.next;
        }
        return true;
    }

    public void PrintList() {
        Node Current = first;
        while (Current != null) {
            System.out.println(Current.item);
            Current = Current.next;
        }
    }
}

class Node {
    public int item;
    public Node next;

    public Node() {
    }

    public Node(int item) {
        this.item = item;
    }

}